pkg/kubelet/logs/container_log_manager.go

```diff
@@ -0,0 +1,401 @@
+	// TODO(#59998): Use kubelet pod cache.
+	containers, err := c.runtimeService.ListContainers(&runtimeapi.ContainerFilter{})
+	if err != nil {
+		return fmt.Errorf("failed to list containers: %v", err)
+	}
+	// NOTE(random-liu): Figure out whether we need to rotate container logs in parallel.
+	for _, container := range containers {
+		// Only rotate logs for running containers. Non-running containers won't
+		// generate new output, it doesn't make sense to keep an empty latest log.
+		if container.GetState() != runtimeapi.ContainerState_CONTAINER_RUNNING {
+			continue
+		}
+		id := container.GetId()
+		// Note that we should not block log rotate for an error of a single container.
+		status, err := c.runtimeService.ContainerStatus(id)
+		if err != nil {
+			glog.Errorf("Failed to get container status for %q: %v", id, err)
+			continue
+		}
+		path := status.GetLogPath()
+		info, err := os.Stat(path)
+		if err != nil {
+			if !os.IsNotExist(err) {
+				glog.Errorf("Failed to stat container log %q: %v", path, err)
+				continue
+			}
+			// In rotateLatestLog, there are several cases that we may
+			// lose original container log after ReopenContainerLog fails.
+			// We try to to recover it by reopening container log.
+			if err := c.runtimeService.ReopenContainerLog(id); err != nil {
+				glog.Errorf("Container %q log %q doesn't exist, reopen container log failed: %v", id, path, err)
+				continue
+			}
+			// The container log should be recovered.
+			info, err = os.Stat(path)
+			if err != nil {
+				glog.Errorf("Failed to stat container log %q after reopen: %v", path, err)
+				continue
+			}
+		}
+		if info.Size() < c.policy.MaxSize {
+			continue
+		}
+		// Perform log rotation.
+		if err := c.rotateLog(id, path); err != nil {
+			glog.Errorf("Failed to rotate log %q for container %q: %v", path, id, err)
+			continue
+		}
+	}
+	return nil
+}
+
+func (c *containerLogManager) rotateLog(id, log string) error {
+	// pattern is used to match all rotated files.
+	pattern := fmt.Sprintf("%s.*", log)
+	logs, err := filepath.Glob(pattern)
+	if err != nil {
+		return fmt.Errorf("failed to list all log files with pattern %q: %v", pattern, err)
+	}
+
+	logs, err = c.cleanupUnusedLogs(logs)
+	if err != nil {
+		return fmt.Errorf("failed to cleanup logs: %v", err)
+	}
+
+	logs, err = c.removeExcessLogs(logs)
+	if err != nil {
+		return fmt.Errorf("failed to remove excess logs: %v", err)
+	}
+
+	// Compress uncompressed log files.
+	for _, l := range logs {
+		if strings.HasSuffix(l, compressSuffix) {
+			continue
+		}
+		if err := c.compressLog(l); err != nil {
+			return fmt.Errorf("failed to compress log %q: %v", l, err)
+		}
+	}
+
+	if err := c.rotateLatestLog(id, log); err != nil {
+		return fmt.Errorf("failed to rotate log %q: %v", log, err)
+	}
+
+	return nil
+}
+
+// cleanupUnusedLogs cleans up temporary or unused log files generated by previous log rotation
+// failure.
+// logs passed in must be in increasing order.
```

Comment:
I must be missing something here.  Why do logs need to be sorted?

Reply:
Increasing order with respect to time?  If so, can you instead say something like `must be ordered from oldest to newest` (or vice-versa if I am mixed up).

Reply:
That's left from my previous implementation. Removed the comment.